extensions/iam/distributed-identity/identity-did-web/src/main/java/org/eclipse/dataspaceconnector/iam/did/web/resolution/WebDidResolver.java

```diff
@@ -0,0 +1,133 @@
+/*
+ *  Copyright (c) 2021 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+package org.eclipse.dataspaceconnector.iam.did.web.resolution;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.dnsoverhttps.DnsOverHttps;
+import org.eclipse.dataspaceconnector.iam.did.spi.document.DidDocument;
+import org.eclipse.dataspaceconnector.iam.did.spi.resolution.DidResolver;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Resolves a Web DID according to the Web DID specification (https://w3c-ccg.github.io/did-method-web).
+ */
+public class WebDidResolver implements DidResolver {
+    private static final String DID_SCHEME = "did";
+    private static final String DID_METHOD = "web";
+    private static final String DID_METHOD_PREFIX = DID_METHOD + ":";
+    private static final String HTTPS_PREFIX = " https://";
+    private static final String DID_DOCUMENT = "did.json";
+    private static final String WELL_KNOWN = "/.well-known/";
+
+    private OkHttpClient httpClient;
+    private ObjectMapper mapper;
+    private Monitor monitor;
+
+    /**
+     * Creates a resolver that executes DNS over HTTPs queries against the given DNS server.
+     */
+    public WebDidResolver(@Nullable URL dnsServer, OkHttpClient httpClient, ObjectMapper mapper, Monitor monitor) {
+        if (dnsServer != null) {
+            // use DNS over HTTPS for name lookups
+            var dns = new DnsOverHttps.Builder().client(httpClient).url(requireNonNull(HttpUrl.get(dnsServer))).includeIPv6(false).build();
+            this.httpClient = httpClient.newBuilder().dns(dns).build();
+        } else {
+            // use standard DNS lookups
+            this.httpClient = httpClient;
+        }
+        this.mapper = mapper;
+        this.monitor = monitor;
+    }
+
+    /**
+     * Creates a resolver that executes standard DNS lookups.
+     */
+    public WebDidResolver(OkHttpClient httpClient, ObjectMapper mapper, Monitor monitor) {
+        this(null, httpClient, mapper, monitor);
+    }
+
+    @Override
+    public @NotNull String getMethod() {
+        return DID_METHOD;
+    }
+
+    @Override
+    @Nullable
+    public DidDocument resolve(String didKey) {
```

Comment:
I think we should define when a method have to return an object/null or a `Result` instance, because the `DidPublicKeyResolver` returns a `Result` object, and this could also be a case to use that pattern, to improve the error message returned by `DidResolverRegistryImpl`

Reply:
This is another really good idea and simplification, I added it in.